File: Src/Lib/Shared/P32.lua
-- Engine script for the P32AC-DM operated by Amtrak and Metro-North.
--
-- @include RollingStock/PowerSupply/Electrification.lua
-- @include RollingStock/PowerSupply/PowerSupply.lua
-- @include RollingStock/BrakeLight.lua
-- @include SafetySystems/AspectDisplay/Genesis.lua
-- @include SafetySystems/Alerter.lua
-- @include Flash.lua
-- @include Iterator.lua
-- @include Misc.lua
-- @include RailWorks.lua
-- @include Scheduler.lua
-- @include Units.lua
local powermode = {thirdrail = 0, diesel = 1}

local playersched, anysched
local adu
local alerter
local power
local blight
local ditchflasher
local state = {
  throttle = 0,
  train_brake = 0,
  acknowledge = false,

  speed_mps = 0,
  acceleration_mps2 = 0,
  trackspeed_mps = 0,
  consistlength_m = 0,
  speedlimits = {},
  restrictsignals = {},

  lastchangetime_s = nil,
  lasthorntime_s = nil
}

Initialise = Misc.wraperrors(function()
  playersched = Scheduler:new{}
  anysched = Scheduler:new{}

  adu = GenesisAdu:new{
    isamtrak = isamtrak,
    getbrakesuppression = function() return state.train_brake >= 0.4 end,
    getacknowledge = function() return state.acknowledge end,
    getspeed_mps = function() return state.speed_mps end,
    gettrackspeed_mps = function() return state.trackspeed_mps end,
    getconsistlength_m = function() return state.consistlength_m end,
    iterspeedlimits = function() return pairs(state.speedlimits) end,
    iterrestrictsignals = function() return pairs(state.restrictsignals) end
  }

  alerter = Alerter:new{}
  alerter:start()

  local iselectric = string.sub(RailWorks.GetRVNumber(), 1, 1) == "T"
  power = PowerSupply:new{
    modecontrol = "PowerMode",
    -- The power mode control is reversed in DTG's cab car; compensate for this
    -- (while sacrificing P32-to-P32 MU capability).
    modereadfn = function(v)
      local ishelper = not RailWorks.GetIsEngineWithKey() and
                         RailWorks.GetIsPlayer()
      local mode = ishelper and 1 - v or v
      -- AI locos require some extra safeguards. Control values may be raw.
      return math.max(Misc.round(mode), 0)
    end,
    transition_s = 20,
    getcantransition = function() return state.throttle <= 0 end,
    modes = {
      [powermode.thirdrail] = function(elec)
        return elec:isavailable(Electrification.type.thirdrail)
      end,
      [powermode.diesel] = function(elec) return true end
    },
    getautomode = function(cp)
      if cp == Electrification.autochangepoint.ai_to_thirdrail then
        return powermode.thirdrail
      elseif cp == Electrification.autochangepoint.ai_to_diesel then
        return powermode.diesel
      else
        return nil
      end
    end
  }
  power:setavailable(Electrification.type.thirdrail, iselectric)

  blight = BrakeLight:new{}

  local ditchflash_s = 0.65
  ditchflasher = Flash:new{
    scheduler = playersched,
    off_s = ditchflash_s,
    on_s = ditchflash_s
  }

  RailWorks.BeginUpdate()
end)

local function readcontrols()
  state.throttle = RailWorks.GetControlValue("VirtualThrottle", 0)
  state.train_brake = RailWorks.GetControlValue("TrainBrakeControl", 0)
  state.acknowledge = RailWorks.GetControlValue("AWSReset", 0) > 0
  if state.acknowledge then alerter:acknowledge() end

  if RailWorks.GetControlValue("Horn", 0) > 0 then
    state.lasthorntime_s = playersched:clock()
  end
end

local function readlocostate()
  state.speed_mps = RailWorks.GetControlValue("SpeedometerMPH", 0) *
                      Units.mph.tomps
  state.acceleration_mps2 = RailWorks.GetAcceleration()
  state.trackspeed_mps = RailWorks.GetCurrentSpeedLimit(1)
  state.consistlength_m = RailWorks.GetConsistLength()
  state.speedlimits = Iterator.totable(Misc.iterspeedlimits(
                                         Acses.nlimitlookahead))
  state.restrictsignals = Iterator.totable(
                            Misc.iterrestrictsignals(Acses.nsignallookahead))
end

local function writelocostate()
  local penalty = alerter:ispenalty() or adu:ispenalty()
  local haspower = power:haspower()
  local throttle = (penalty or not haspower) and 0 or state.throttle
  RailWorks.SetControlValue("Regulator", 0, throttle)
  RailWorks.SetPowerProportion(-1, Misc.intbool(haspower))
  -- There's no virtual train brake, so just move the braking handle.
  if penalty then RailWorks.SetControlValue("TrainBrakeControl", 0, 0.85) end

  -- DTG's "blended braking" algorithm
  local dynbrake
  local pipepress_psi = 70 -
                          RailWorks.GetControlValue("AirBrakePipePressurePSI", 0)
  if power:getmode() == powermode.thirdrail then
    dynbrake = 0
  elseif pipepress_psi > 0 then
    dynbrake = pipepress_psi * 0.01428
  else
    dynbrake = 0
  end
  RailWorks.SetControlValue("DynamicBrake", 0, dynbrake)

  local alarm = alerter:isalarm() or adu:isalarm()
  local alert = adu:isalertplaying()
  RailWorks.SetControlValue("AWS", 0, Misc.intbool(alarm or alert))
  RailWorks.SetControlValue("AWSWarnCount", 0, Misc.intbool(alarm))
end

local function setnonplayerstate()
  -- Read throttle and power mode from the lead locomotive.
  state.throttle = RailWorks.GetControlValue("Regulator", 0)
  RailWorks.SetPowerProportion(-1, Misc.intbool(power:haspower()))
end

local function setcutin()
  if not playersched:isstartup() then
    adu:setatcstate(RailWorks.GetControlValue("ATCCutIn", 0) == 1)
    adu:setacsesstate(RailWorks.GetControlValue("ACSESCutIn", 0) == 1)
  end
end

local function setadu()
  local aspect = adu:getaspect()
  local n, l, m, r
  if aspect == GenesisAdu.aspect.restrict then
    n, l, m, r = 0, 0, 0, 1
  elseif aspect == GenesisAdu.aspect.medium then
    n, l, m, r = 0, 0, 1, 0
  elseif aspect == GenesisAdu.aspect.limited then
    n, l, m, r = 0, 1, 0, 0
  elseif aspect == GenesisAdu.aspect.clear then
    n, l, m, r = 1, 0, 0, 0
  end
  RailWorks.SetControlValue("SigN", 0, n)
  RailWorks.SetControlValue("SigL", 0, l)
  RailWorks.SetControlValue("SigM", 0, m)
  RailWorks.SetControlValue("SigR", 0, r)

  local sigspeed_mph = adu:getsignalspeed_mph()
  if sigspeed_mph == nil then
    RailWorks.SetControlValue("SignalSpeed", 0, 1) -- hide
  else
    RailWorks.SetControlValue("SignalSpeed", 0, sigspeed_mph)
  end
end

local function setdisplay()
  local speed_mph = RailWorks.GetControlValue("SpeedometerMPH", 0)
  RailWorks.SetControlValue("SpeedoHundreds", 0, Misc.getdigit(speed_mph, 2))
  RailWorks.SetControlValue("SpeedoTens", 0, Misc.getdigit(speed_mph, 1))
  RailWorks.SetControlValue("SpeedoUnits", 0, Misc.getdigit(speed_mph, 0))
  RailWorks.SetControlValue("SpeedoDecimal", 0, Misc.getdigit(speed_mph, -1))

  local overspeed_mph = adu:getoverspeed_mph()
  if overspeed_mph == nil then
    RailWorks.SetControlValue("TrackHundreds", 0, -1)
    RailWorks.SetControlValue("TrackTens", 0, -1)
    RailWorks.SetControlValue("TrackUnits", 0, -1)
  else
    RailWorks.SetControlValue("TrackHundreds", 0,
                              Misc.getdigit(overspeed_mph, 2))
    RailWorks.SetControlValue("TrackTens", 0, Misc.getdigit(overspeed_mph, 1))
    RailWorks.SetControlValue("TrackUnits", 0, Misc.getdigit(overspeed_mph, 0))
  end

  RailWorks.SetControlValue("AlerterVisual", 0, Misc.intbool(alerter:isalarm()))
end

local function setditchlights()
  local horntime_s = 30
  local horn = state.lasthorntime_s ~= nil and playersched:clock() <=
                 state.lasthorntime_s + horntime_s
  local flash = isamtrak and horn
  local headlights = RailWorks.GetControlValue("Headlights", 0)
  local crosslights = RailWorks.GetControlValue("CrossingLight", 0) == 1
  local fixed = headlights > 0.5 and headlights < 1.5 and crosslights and
                  not flash
  ditchflasher:setflashstate(flash)
  local flashleft = ditchflasher:ison()

  local showleft = fixed or (flash and flashleft)
  RailWorks.ActivateNode("ditch_left", showleft)
  Call("DitchLight_L:Activate", Misc.intbool(showleft))

  local showright = fixed or (flash and not flashleft)
  RailWorks.ActivateNode("ditch_right", showright)
  Call("DitchLight_R:Activate", Misc.intbool(showright))
end

local function setcablights()
  local function activate(v) return Misc.intbool(v > 0.8) end
  -- engineer's side task light
  Call("CabLight_R:Activate", activate(RailWorks.GetControlValue("CabLight", 0)))
  -- engineer's forward task light
  Call("TaskLight_R:Activate",
       activate(RailWorks.GetControlValue("CabLight1", 0)))
  -- secondman's forward task light
  Call("TaskLight_L:Activate",
       activate(RailWorks.GetControlValue("CabLight2", 0)))
  -- secondman's side task light
  Call("CabLight_L:Activate",
       activate(RailWorks.GetControlValue("CabLight4", 0)))
  -- dome light
  Call("CabLight_M:Activate",
       activate(RailWorks.GetControlValue("CabLight5", 0)))
end

local function setexhaust()
  local r, g, b, rate
  local minrpm = 180
  local effort = RailWorks.GetTractiveEffort()
  if power:getmode() == powermode.thirdrail or
    RailWorks.GetControlValue("RPM", 0) < minrpm then
    r, g, b = 0, 0, 0
    rate = 0
    -- DTG's exhaust logic
  elseif effort < 0.1 then
    r, g, b = 0.25, 0.25, 0.25
    rate = 0.01
  elseif effort >= 0.1 and effort < 0.5 then
    r, g, b = 0.1, 0.1, 0.1
    rate = 0.005
  else
    r, g, b = 0, 0, 0
    rate = 0.001
  end
  Call("DieselExhaust:SetEmitterColour", r, g, b)
  Call("DieselExhaust:SetEmitterRate", rate)
end

local function updateplayer(dt)
  readcontrols()
  readlocostate()

  playersched:update()
  anysched:update()
  adu:update(dt)
  alerter:update(dt)
  power:update(dt)
  blight:playerupdate()

  writelocostate()
  setcutin()
  setadu()
  setdisplay()
  setditchlights()
  setcablights()
  setexhaust()
end

local function updatenonplayer(dt)
  anysched:update()
  power:update(dt)

  setnonplayerstate()
  setditchlights()
  setcablights()
  setexhaust()
end

Update = Misc.wraperrors(function(dt)
  if RailWorks.GetIsEngineWithKey() then
    updateplayer(dt)
  else
    updatenonplayer(dt)
  end
end)

OnControlValueChange = Misc.wraperrors(function(name, index, value)
  if name == "ExpertPowerMode" and RailWorks.GetIsEngineWithKey() and
    not anysched:isstartup() and (value == 0 or value == 1) then
    Misc.showalert("Not available in OpenNEC")
  end

  if name == "VirtualThrottle" or name == "TrainBrakeControl" then
    alerter:acknowledge()
  end

  RailWorks.SetControlValue(name, index, value)
end)

OnCustomSignalMessage = Misc.wraperrors(function(message)
  power:receivemessage(message)
  adu:receivemessage(message)
end)

OnConsistMessage = Misc.wraperrors(function(message, argument, direction)
  blight:receivemessage(message, argument, direction)

  RailWorks.Engine_SendConsistMessage(message, argument, direction)
end)

File: src/lib/shared/p32.ts
/**
 * Metro-North/Amtrak GE P32AC-DM
 */

import * as ale from "lib/alerter";
import * as c from "lib/constants";
import * as frp from "lib/frp";
import { FrpEngine } from "lib/frp-engine";
import { mapBehavior, rejectRepeats } from "lib/frp-extra";
import { SensedDirection } from "lib/frp-vehicle";
import * as m from "lib/math";
import { AduAspect } from "lib/nec/adu";
import * as cs from "lib/nec/cabsignals";
import * as adu from "lib/nec/twospeed-adu";
import * as ps from "lib/power-supply";
import * as rw from "lib/railworks";
import * as fx from "lib/special-fx";
import * as ui from "lib/ui";

export const dualModeOrder: [ps.EngineMode.ThirdRail, ps.EngineMode.Diesel] = [
    ps.EngineMode.ThirdRail,
    ps.EngineMode.Diesel,
];
export const dualModeSwitchS = 20;

const ditchLightsFadeS = 0.3;
const ditchLightFlashS = 0.65;
const taskLightFadeS = 0.8 / 2;

export function onInit(me: FrpEngine, isAmtrak: boolean) {
    // Dual-mode power supply
    const electrification = ps.createElectrificationBehaviorWithControlValues(me, {
        [ps.Electrification.Overhead]: "PowerOverhead",
        [ps.Electrification.ThirdRail]: "Power3rdRail",
    });
    const modeAuto = () => (me.rv.GetControlValue("ExpertPowerMode") as number) > 0.5;
    ui.createAutoPowerStatusPopup(me, modeAuto);
    const modeAutoSwitch$ = ps.createDualModeAutoSwitchStream(me, ...dualModeOrder, modeAuto);
    modeAutoSwitch$(mode => {
        me.rv.SetControlValue("PowerMode", mode === ps.EngineMode.Diesel ? 1 : 0);
    });
    const modeSelect = () => {
        const cv = Math.round(me.rv.GetControlValue("PowerMode") as number);
        return cv < 0.5 ? ps.EngineMode.ThirdRail : ps.EngineMode.Diesel;
    };
    const modePosition = ps.createDualModeEngineBehavior({
        e: me,
        modes: dualModeOrder,
        getPlayerMode: modeSelect,
        getAiMode: modeSelect,
        getPlayerCanSwitch: () => {
            const throttle = me.rv.GetControlValue("VirtualThrottle") as number;
            return throttle < 0.5;
        },
        transitionS: dualModeSwitchS,
        instantSwitch: modeAutoSwitch$,
        positionFromSaveOrConsist: () => (me.rv.GetControlValue("PowerStart") as number) - 1,
    });
    const setModePosition$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(modePosition),
        rejectRepeats()
    );
    setModePosition$(position => {
        me.rv.SetControlValue("PowerStart", position + 1);
    });
    // Power3rdRail is not set correctly in the third-rail engine blueprint, so
    // set it ourselves based on the value of PowerMode.
    const fixElectrification$ = frp.compose(
        me.createFirstUpdateAfterControlsSettledStream(),
        frp.filter(resumeFromSave => !resumeFromSave),
        mapBehavior(modeSelect),
        frp.map(mode => (mode === ps.EngineMode.ThirdRail ? 1 : 0))
    );
    fixElectrification$(thirdRail => {
        me.rv.SetControlValue("Power3rdRail", thirdRail);
    });

    // Safety systems cut in/out
    const atcCutIn = () => (me.rv.GetControlValue("ATCCutIn") as number) > 0.5;
    const acsesCutIn = () => (me.rv.GetControlValue("ACSESCutIn") as number) > 0.5;
    ui.createAtcStatusPopup(me, atcCutIn);
    ui.createAcsesStatusPopup(me, acsesCutIn);
    const alerterCutIn = frp.liftN((atcCutIn, acsesCutIn) => atcCutIn || acsesCutIn, atcCutIn, acsesCutIn);
    ui.createAlerterStatusPopup(me, alerterCutIn);

    // Safety systems and ADU
    const acknowledge = me.createAcknowledgeBehavior();
    const suppression = () => (me.rv.GetControlValue("TrainBrakeControl") as number) >= 0.4;
    const [aduState$, aduEvents$] = adu.create({
        atc: isAmtrak ? cs.fourAspectAtc : cs.metroNorthAtc,
        e: me,
        acknowledge,
        suppression,
        atcCutIn,
        acsesCutIn,
        equipmentSpeedMps: (isAmtrak ? 110 : 80) * c.mph.toMps,
        pulseCodeControlValue: "SignalSpeedLimit",
    });
    const aduStateHub$ = frp.compose(aduState$, frp.hub());
    aduStateHub$(({ aspect, trackSpeedMph }) => {
        me.rv.SetControlValue("SigN", aspect === cs.FourAspect.Clear ? 1 : 0);
        me.rv.SetControlValue("SigL", aspect === cs.FourAspect.ApproachLimited ? 1 : 0);
        me.rv.SetControlValue("SigM", aspect === cs.FourAspect.Approach ? 1 : 0);
        me.rv.SetControlValue("SigR", aspect === cs.FourAspect.Restricting || aspect === AduAspect.Stop ? 1 : 0);

        me.rv.SetControlValue(
            "SignalSpeed",
            {
                [AduAspect.Stop]: 0,
                [cs.FourAspect.Restricting]: 15,
                [cs.FourAspect.Approach]: 30,
                [cs.FourAspect.ApproachLimited]: 45,
                [cs.FourAspect.Clear]: -1,
            }[aspect]
        );

        if (trackSpeedMph !== undefined) {
            const [[h, t, u]] = m.digits(trackSpeedMph, 3);
            me.rv.SetControlValue("TrackHundreds", h);
            me.rv.SetControlValue("TrackTens", t);
            me.rv.SetControlValue("TrackUnits", u);
        } else {
            me.rv.SetControlValue("TrackHundreds", -1);
            me.rv.SetControlValue("TrackTens", -1);
            me.rv.SetControlValue("TrackUnits", -1);
        }
    });
    const aduState = frp.stepper(aduStateHub$, undefined);
    // Alerter
    const alerter$ = frp.compose(ale.create({ e: me, acknowledge, cutIn: alerterCutIn }), frp.hub());
    const alerterState = frp.stepper(alerter$, undefined);
    alerter$(({ alarm }) => {
        me.rv.SetControlValue("AlerterVisual", alarm ? 1 : 0);
    });
    // Safety system sounds
    const upgradeEvents$ = frp.compose(
        aduEvents$,
        frp.filter(evt => evt === adu.AduEvent.Upgrade)
    );
    const upgradeSound$ = frp.compose(me.createPlayerWithKeyUpdateStream(), fx.triggerSound(1, upgradeEvents$));
    const alarmsUpdate$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(
            frp.liftN(
                (aduState, alerterState, upgradeSound) => {
                    return {
                        aws: aduState?.atcAlarm || aduState?.acsesAlarm || alerterState?.alarm || upgradeSound,
                        awsWarnCount: (aduState?.atcAlarm || aduState?.acsesAlarm || alerterState?.alarm) ?? false,
                    };
                },
                aduState,
                alerterState,
                frp.stepper(upgradeSound$, false)
            )
        )
    );
    alarmsUpdate$(({ aws, awsWarnCount }) => {
        me.rv.SetControlValue("AWS", aws ? 1 : 0);
        me.rv.SetControlValue("AWSWarnCount", awsWarnCount ? 1 : 0);
    });

    // Throttle, dynamic brake, and air brake controls
    const isPenaltyBrake = frp.liftN(
        (aduState, alerterState) => (aduState?.penaltyBrake || alerterState?.penaltyBrake) ?? false,
        aduState,
        alerterState
    );
    const isPowerAvailable = frp.liftN(
        position => ps.dualModeEngineHasPower(position, ...dualModeOrder, electrification),
        modePosition
    );
    const throttle = frp.liftN(
        (isPenaltyBrake, isPowerAvailable, input) => (isPenaltyBrake || !isPowerAvailable ? 0 : input),
        isPenaltyBrake,
        isPowerAvailable,
        () => me.rv.GetControlValue("VirtualThrottle") as number
    );
    const throttle$ = frp.compose(me.createPlayerWithKeyUpdateStream(), mapBehavior(throttle));
    throttle$(v => {
        me.rv.SetControlValue("Regulator", v);
    });
    // There's no virtual train brake, so just move the braking handle.
    const fullService = 0.85;
    const setPenaltyBrake$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        frp.filter(_ => frp.snapshot(isPenaltyBrake)),
        frp.map(_ => fullService)
    );
    setPenaltyBrake$(v => {
        me.rv.SetControlValue("TrainBrakeControl", v);
    });
    // DTG's "blended braking" algorithm
    const dynamicBrake = frp.liftN(
        (modePosition, brakePipePsi) => {
            const isDiesel = modePosition === 1;
            return isDiesel ? (70 - brakePipePsi) * 0.01428 : 0;
        },
        modePosition,
        () => me.rv.GetControlValue("AirBrakePipePressurePSI") as number
    );
    const dynamicBrake$ = frp.compose(me.createPlayerWithKeyUpdateStream(), mapBehavior(dynamicBrake));
    dynamicBrake$(v => {
        me.rv.SetControlValue("DynamicBrake", v);
    });

    // Ditch lights
    const ditchLights = [
        new fx.FadeableLight(me, ditchLightsFadeS, "DitchLight_L"),
        new fx.FadeableLight(me, ditchLightsFadeS, "DitchLight_R"),
    ];
    const areHeadLightsOn = () => {
        const cv = me.rv.GetControlValue("Headlights") as number;
        return cv > 0.5 && cv < 1.5;
    };
    const areDitchLightsOn = frp.liftN(
        (headLights, crossingLight) => headLights && crossingLight,
        areHeadLightsOn,
        () => (me.rv.GetControlValue("CrossingLight") as number) > 0.5
    );
    const hornSequencer$ = frp.compose(me.createPlayerWithKeyUpdateStream(), me.mapGetCvStream("CylinderCock"));
    const ditchLightsHornTimeS$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        me.mapGetCvStream("Horn"),
        frp.merge(hornSequencer$),
        frp.filter(v => v === 1),
        frp.map(_ => me.e.GetSimulationTime())
    );
    const ditchLightsFlash = frp.liftN(
        (ditchOn, bell, hornTimeS) =>
            isAmtrak && ditchOn && (bell || (hornTimeS !== undefined && me.e.GetSimulationTime() - hornTimeS <= 30)),
        areDitchLightsOn,
        () => (me.rv.GetControlValue("Bell") as number) > 0.5,
        frp.stepper(ditchLightsHornTimeS$, undefined)
    );
    const ditchLightsPlayer$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        fx.behaviorStopwatchS(ditchLightsFlash),
        frp.map((stopwatchS): [boolean, boolean] => {
            if (stopwatchS === undefined) {
                const ditchOn = frp.snapshot(areDitchLightsOn);
                return [ditchOn, ditchOn];
            } else {
                const flashLeft = stopwatchS % (ditchLightFlashS * 2) < ditchLightFlashS;
                return [flashLeft, !flashLeft];
            }
        })
    );
    const ditchLightsAi$ = frp.compose(
        me.createAiUpdateStream(),
        frp.map((au): [boolean, boolean] => {
            const ditchOn = au.direction === SensedDirection.Forward;
            return [ditchOn, ditchOn];
        })
    );
    const ditchLights$ = frp.compose(
        me.createPlayerWithoutKeyUpdateStream(),
        frp.map((_): [boolean, boolean] => [false, false]),
        frp.merge(ditchLightsPlayer$),
        frp.merge(ditchLightsAi$)
    );
    ditchLights$(([l, r]) => {
        const [left, right] = ditchLights;
        left.setOnOff(l);
        right.setOnOff(r);
    });
    const ditchNodeLeft$ = frp.compose(
        me.createUpdateStream(),
        frp.map(_ => {
            const [light] = ditchLights;
            return light.getIntensity() > 0.5;
        }),
        rejectRepeats()
    );
    const ditchNodeRight$ = frp.compose(
        me.createUpdateStream(),
        frp.map(_ => {
            const [, light] = ditchLights;
            return light.getIntensity() > 0.5;
        }),
        rejectRepeats()
    );
    ditchNodeLeft$(on => {
        me.rv.ActivateNode("ditch_left", on);
    });
    ditchNodeRight$(on => {
        me.rv.ActivateNode("ditch_right", on);
    });

    // Horn rings the bell.
    const bellControl$ = frp.compose(
        me.createOnCvChangeStreamFor("Horn"),
        frp.merge(me.createOnCvChangeStreamFor("CylinderCock")),
        frp.filter(v => v === 1),
        me.mapAutoBellStream()
    );
    bellControl$(v => {
        me.rv.SetControlValue("Bell", v);
    });

    // Diesel exhaust effect
    const exhaust = new rw.Emitter("DieselExhaust");
    const exhaustState = frp.liftN(
        (modePosition, rpm, effort): [color: number, rate: number] => {
            // DTG's exhaust logic
            if (modePosition < 0.5) {
                return [0, 0];
            } else if (rpm < 180) {
                return [0, 0];
            } else if (effort < 0.1) {
                return [0.25, 0.01];
            } else if (effort < 0.5) {
                return [0.1, 0.005];
            } else {
                return [0, 0.001];
            }
        },
        modePosition,
        () => me.rv.GetControlValue("RPM") as number,
        () => me.eng.GetTractiveEffort()
    );
    const exhaustActive$ = frp.compose(
        me.createUpdateStream(),
        mapBehavior(exhaustState),
        frp.map(([, rate]) => rate > 0),
        rejectRepeats()
    );
    const exhaustColor$ = frp.compose(
        me.createUpdateStream(),
        mapBehavior(exhaustState),
        frp.map(([color]) => color),
        rejectRepeats()
    );
    const exhaustRate$ = frp.compose(
        me.createUpdateStream(),
        mapBehavior(exhaustState),
        frp.map(([, rate]) => rate),
        rejectRepeats()
    );
    exhaustActive$(active => {
        exhaust.SetEmitterActive(active);
    });
    exhaustColor$(color => {
        exhaust.SetEmitterColour(color, color, color);
    });
    exhaustRate$(rate => {
        exhaust.SetEmitterRate(rate);
    });

    // Speedometer
    const aSpeedoMph$ = frp.compose(
        me.createPlayerWithKeyUpdateStream(),
        mapBehavior(me.createSpeedometerMpsBehavior()),
        frp.map(mps => mps * c.mps.toMph),
        frp.map(mph => Math.abs(mph))
    );
    aSpeedoMph$(mph => {
        const [[h, t, u]] = m.digits(Math.floor(mph), 3);
        me.rv.SetControlValue("SpeedoHundreds", h);
        me.rv.SetControlValue("SpeedoTens", t);
        me.rv.SetControlValue("SpeedoUnits", u);
        me.rv.SetControlValue("SpeedoDecimal", m.tenths(mph));
    });

    // Cockpit lights
    // Engineer's side task light
    createTaskLight(me, "CabLight_R", "CabLight");
    // Engineer's forward task light
    createTaskLight(me, "TaskLight_R", "CabLight1");
    // Secondman's forward task light
    createTaskLight(me, "TaskLight_L", "CabLight2");
    // Secondman's side task light
    createTaskLight(me, "CabLight_L", "CabLight4");
    // Dome light
    createTaskLight(me, "CabLight_M", "CabLight5");

    // Process OnControlValueChange events.
    const onCvChange$ = frp.compose(
        me.createOnCvChangeStream(),
        frp.reject(([name]) => name === "Bell")
    );
    onCvChange$(([name, value]) => {
        me.rv.SetControlValue(name, value);
    });

    // Set consist brake lights.
    fx.createBrakeLightStreamForEngine(me);

    // Set platform door height.
    fx.setLowPlatformDoorsForEngine(me, true);

    // Enable updates.
    me.e.BeginUpdate();
}

function createTaskLight(e: FrpEngine, light: string, cv: string) {
    const fadeableLight = new fx.FadeableLight(e, taskLightFadeS, light);
    const stream$ = frp.compose(
        e.createPlayerWithKeyUpdateStream(),
        e.mapGetCvStream(cv),
        frp.map(v => v > 0.5)
    );
    stream$(on => {
        fadeableLight.setOnOff(on);
    });
}
